icu4c/source/test/intltest/ucdtest.cpp

```diff
@@ -1024,3 +1027,107 @@ void UnicodeTest::TestIDCompatMath() {
     assertTrue("idcmStart.contains(U+1D7C3)", idcmStart.contains(0x1D7C3));
     assertFalse("idcmStart.contains(U+1D7C4)", idcmStart.contains(0x1D7C4));
 }
+
+
+U_NAMESPACE_BEGIN
+
+class BuiltInPropertyNames : public PropertyNames {
+public:
+    ~BuiltInPropertyNames() {}
+
+    int32_t getPropertyEnum(const char *name) const override {
+      return (int32_t) u_getPropertyEnum(name);
+    }
+
+    int32_t getPropertyValueEnum(int32_t property, const char *name) const override {
+        return (int32_t) u_getPropertyValueEnum((UProperty) property, name);
+    }
+};
+
+U_NAMESPACE_END
+
+void UnicodeTest::TestBinaryPropertyUsingPpucd() {
+    IcuTestErrorCode errorCode(*this, "TestIDCompatMathStartPpucd()");
+
+    // Iniitalize PPUCD parsing object using file in repo and using
+    // property names present in built-in data in ICU
+    char path[500];
+    // get path to `source/data/unidata/` including trailing `/`
+    char *unidataPath = getUnidataPath(path);
+    if(unidataPath == nullptr) {
+        fprintf(stderr, "TestBinaryPropertyUsingPpucd: unable to open ppucd.txt from ICU source tree\n");
+        errln("exiting early because unable to open ppucd.txt from ICU source tree");
+    }
+    char *unidataPathEnd = strchr(path, 0);
+    strcpy(unidataPathEnd, "ppucd.txt");
+    CharString ppucdPath(path, errorCode);
+    PreparsedUCD ppucd(ppucdPath.data(), errorCode);
+    if(errorCode.isFailure()) {
+        fprintf(stderr, "TestBinaryPropertyUsingPpucd: unable to open %s - %s\n",
+                ppucdPath.data(), errorCode.errorName());
+        errln("unable to open %s - %s\n",
+                ppucdPath.data(), errorCode.errorName());
+    }
+    BuiltInPropertyNames builtInPropNames;
+    ppucd.setPropertyNames(&builtInPropNames);
+
+    // Define which binary properties we want to compare
+    UProperty propsUnderTest[] = {
+        UCHAR_ID_COMPAT_MATH_START,
+        UCHAR_ID_COMPAT_MATH_CONTINUE,
+    };
+    uint32_t numProps = 2;
+
+    // Allocate & initialize UnicodeSets per binary property from PPUCD data
+    UnicodeSet *ppucdPropSets = new UnicodeSet[numProps];
+
+    // Iterate through PPUCD file, accumulating each line's data into each UnicodeSet per property
+    PreparsedUCD::LineType lineType;
+    UnicodeSet newValues;
+    while((lineType=ppucd.readLine(errorCode))!=PreparsedUCD::NO_LINE) {
+        if(ppucd.lineHasPropertyValues()) {
+            const UniProps *lineProps=ppucd.getProps(newValues, errorCode);
+
+            for(uint32_t i = 0; i < numProps; i++) {
+                UProperty prop = propsUnderTest[i];
+                int32_t propEnumVal = (int) prop;
+                // TODO: add or remove ranges according to value
+                if (lineProps->binProps[(propEnumVal)]) {
```

Comment:
Before you check the value, you should check if the line has data for the property. For example:
```
if (!newValues.contains(prop)) { continue; }
```